src/plugins.ts

```diff
@@ -0,0 +1,190 @@
+import * as fs from 'mz/fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+import { Logger, NoopLogger } from './logging';
+import { combinePaths } from './match-files';
+import { PluginSettings } from './request-type';
+import { toUnixPath } from './util';
+
+// Based on types and logic from TypeScript server/project.ts @
+// https://github.com/Microsoft/TypeScript/blob/711e890e59e10aa05a43cb938474a3d9c2270429/src/server/project.ts
+
+/**
+ * A plugin exports an initialization function, injected with
+ * the current typescript instance
+ */
+export type PluginModuleFactory = (mod: { typescript: typeof ts }) => PluginModule;
+
+export type EnableProxyFunc = (pluginModuleFactory: PluginModuleFactory, pluginConfigEntry: ts.PluginImport) => void;
+
+/**
+ * A plugin presents this API when initialized
+ */
+export interface PluginModule {
+	create(createInfo: PluginCreateInfo): ts.LanguageService;
+	getExternalFiles?(proj: Project): string[];
+}
+
+/**
+ * All of tsserver's environment exposed to plugins
+ */
+export interface PluginCreateInfo {
+	project: Project;
+	languageService: ts.LanguageService;
+	languageServiceHost: ts.LanguageServiceHost;
+	serverHost: ServerHost;
+	config: any;
+}
+
+/**
+ * The portion of tsserver's Project API exposed to plugins
+ */
+export interface Project {
+	projectService: {
+		logger: Logger;
+	};
+}
+
+/**
+ * The portion of tsserver's ServerHost API exposed to plugins
+ */
+export type ServerHost = object;
+
+/**
+ * The result of a node require: a module or an error.
+ */
+type RequireResult = { module: {}, error: undefined } | { module: undefined, error: {} };
+
+export class PluginLoader {
+
+	private allowLocalPluginLoads: boolean = false;
+	private globalPlugins: string[] = [];
+	private pluginProbeLocations: string[] = [];
+
+	constructor(
+		private rootFilePath: string,
+		private fs: ts.ModuleResolutionHost,
+		pluginSettings?: PluginSettings,
+		private logger = new NoopLogger(),
+		private resolutionHost = new LocalModuleResolutionHost(),
+		private requireModule: (moduleName: string) => any = require) {
+		if (pluginSettings) {
+			this.allowLocalPluginLoads = pluginSettings.allowLocalPluginLoads || false;
+			this.globalPlugins = pluginSettings.globalPlugins || [];
+			this.pluginProbeLocations = pluginSettings.pluginProbeLocations || [];
+		}
+	}
+
+	public loadPlugins(options: ts.CompilerOptions, applyProxy: EnableProxyFunc) {
+		// Search our peer node_modules, then any globally-specified probe paths
+		// ../../.. to walk from X/node_modules/javascript-typescript-langserver/lib/project-manager.js to X/node_modules/
+		const searchPaths = [combinePaths(__filename, '../../..'), ...this.pluginProbeLocations];
```

Comment:
what is the difference between `combinePaths` and `path.resolve()`?

Reply:
Code is copied from TS project, wanted to keep it as close to original as possible (and we had the function already)

Reply:
Okay, that wasn't clear to me. Maybe change the wording "based on" at the top of the file to "copied from"?